server_spec: Tolerate lack of protocol support

jamessan opened the pull request:
This test started failing in Travis since they disabled IPv6 in the
environment:

```
[ RUN      ] serverstart(), serverstop() parses endpoints correctly: FAIL
...build/neovim/neovim/test/functional/eval/server_spec.lua:83: Expected objects to be the same.
Passed in:
(table) {
  [1] = '127.0.0.1:12345' }
Expected:
(table) {
  [1] = '127.0.0.1:12345'
 *[2] = '::1:12345' }
```

Change all tests to ensure a server was actually started before
expecting it to be returned from `serverlist()`.

jamessan committed:
server_spec: Tolerate lack of protocol support

This test started failing in Travis since they disabled IPv6 in the
environment:

[ RUN      ] serverstart(), serverstop() parses endpoints correctly: FAIL
...build/neovim/neovim/test/functional/eval/server_spec.lua:83: Expected objects to be the same.
Passed in:
(table) {
  [1] = '127.0.0.1:12345' }
Expected:
(table) {
  [1] = '127.0.0.1:12345'
 *[2] = '::1:12345' }

Change all tests to ensure a server was actually started before
expecting it to be returned from serverlist().

bfredl commented:
This also guards against ipv4 failure. But `sockconnect()` test also relies on ipv4 port allocation. Can't we require ipv4 or should that testcase also be guarded (perhaps marked pending)?

jamessan commented:
I don't think it's wise to rely on a specific protocol being available so much as the availability of a supported protocol.  Granted, the lack of IPv6 is the most likely to occur currently, but I can easily see environments where the opposite is true, especially in the future.

bfredl commented:
Ok, `sockconnect()` test cases should probably be marked pending then when their respective protocol is not available.

bfredl commented:
(But while ipv4 network access might not be able/decaying in the not-so-far future I would expect 127.0.0.1 to stick around for a long long time)

jamessan committed:
server_requests_spec: Skip test if no ipv4 stack

jamessan committed:
server_requests_spec: Add ipv6 test

jamessan commented:
> I would expect 127.0.0.1 to stick around for a long long time

I'd expect local ipv6 too (to use `::1`), but it seems Travis has completely disabled the ipv6 stack.

Updated the server_requests_spec to mark the ipv4 test pending if it can't start a server.  Since I was there, also added an ipv6 test.

justinmk commented on test/functional/api/server_requests_spec.lua:
```diff
@@ -304,14 +304,30 @@ describe('server -> client', function()
       local server = spawn(nvim_argv)
       set_session(server)
       local address = funcs.serverstart("127.0.0.1:")
+      if #address == 0 then
```
If serverstart() returns empty for some other reason, this won't catch the problem.

But checking for ipv6 support via the shell is annoying and very platform-dependent. :(

justinmk commented:
We could at least require _at least one of_ ipv4 and ipv6.

justinmk commented on test/functional/eval/server_spec.lua:
```diff
@@ -63,23 +63,33 @@ describe('serverstart(), serverstop()', function()
+      table.insert(expected, v6)
+      funcs.serverstart(v6)  -- exists already; ignore
+    end
     eq(expected, funcs.serverlist())
```
Also:

    eq(true, #expected > 0)
